\chapter{Machine and Deep Learning in DNA sequence classification} \label{sec:ml_dl_dna}

\section{Relevant previous work on DNA classification}\label{sec:previous_work}

\gls{DNA} sequence classification is a critical problem in biomedical research, and since current solutions involve the use of homologies, which is a long and costly process, multiple \gls{ML} techniques were used to successfully complete this task. In recent years, numerous articles and papers were published regarding this classification challenge with both traditional \gls{ML} and \gls{DL} approaches.

% [X]
\citeauthor{Nair2010ANNRepresentation}~\cite{Nair2010ANNRepresentation} proposed a unique technique for organism classification based on a combination of \gls{FCGR} and \gls{DNA}. \gls{CGR} displays \gls{DNA} sequences in a unique way and reveals hidden patterns in them. The frequency of sub-sequences contained in the \gls{DNA} sequence is shown by \gls{FCGR}, which is developed from \gls{CGR}. The taxonomic distribution of Eukaryotic species is broken down into eight groups, and \gls{ANN} is used to classify them.

% [X]
\citeauthor{Rizzo2015AClassification}~\cite{Rizzo2015AClassification} introduced a \gls{DNN} based on spectral sequence representation for \gls{DNA} sequence classification. The framework is evaluated on a dataset of 16S genes, and its results are compared to the General Regression Neural Network as well as Naive Bayes, \gls{RF}, and \gls{SVM} classifiers in terms of accuracy and F1 score. When it came to classifying short sequence fragments of 500 bp, the \gls{DL} technique beat all other classifiers.

% [X]
\citeauthor{Nguyen2016DNANetwork}~\cite{Nguyen2016DNANetwork} developed a new method for classifying \gls{DNA} sequences using a \gls{CNN} while treating them as text input. Because the authors employed one-hot vectors to represent sequences as input to the model, the important position information of each nucleotide in sequences is preserved. The authors investigated the suggested model using 12 \gls{DNA} sequence datasets and found substantial improvements in all of them. Out of the 12 datasets used, 10 of them include DNA sequences that wrap around histone proteins (for example, H3 and H4) and the other datasets are Splice and Promoter datasets. This finding suggests that a \gls{CNN} can be used to handle additional sequence challenges in bioinformatics.

% [X]
\citeauthor{LoBosco2017DeepClassification}~\cite{LoBosco2017DeepClassification} offer two distinct \gls{DL} architectures (\gls{CNN} and \gls{RNN}) for \gls{DNA} sequence classification. For five separate classification tasks, they compare their results using a public data set of \gls{DNA} sequences. Two \gls{DL} architectures were examined in this work for automated classification of bacteria species with no sequence preprocessing procedures. In comparison to a traditional \gls{CNN}, the authors have presented a \gls{LSTM} that utilizes the nucleotide locations in a sequence. The \gls{CNN}s outperform the \gls{LSTM} in the four easiest classification tasks, but their performance deteriorates in the last, when the \gls{LSTM} performs better. %character-level one-hot encoding

% [X]
\citeauthor{Abd-Alhalem2020BacterialLayers}~\cite{Abd-Alhalem2020BacterialLayers} proposed, based on a custom layer, a new technique for classification of bacterial \gls{DNA} sequences. The \gls{FCGR} of \gls{DNA} is employed with a \gls{CNN}. With a proper choice of the frequency k-lengthen words occurrence in \gls{DNA} sequences, the \gls{FCGR} is used as a sequence representation technique. The \gls{DNA} sequence is mapped using \gls{FCGR}, which generates a gene sequence image. Both local and global patterns may be seen in this sequence. For image classification, a pre-trained \gls{CNN} is used. 

% [X]
\citeauthor{Chen2017AClassification}~\cite{Chen2017AClassification} explain the distinction between cell-free DNA and normal DNA. The authors employed a variety of classification models to categorize normal and cfDNA, including \gls{KNN}, \gls{SVM}, and \gls{RF}. In comparison to all other algorithms, the authors determined that the \gls{RF} had the best accuracy. %DNA fragmentation patterns

% [X]
\citeauthor{Helaly2019ConvolutionalStudy}~\cite{Helaly2019ConvolutionalStudy} analyze three of the most current \gls{DL} efforts for taxonomy classification using the 16S rRNA barcode dataset. Three distinct \gls{CNN} architectures are examined, as well as three different feature representations: k-mer spectral representation, \gls{FCGR}, and character-level integer encoding. The most fine-grained classification challenge showed that representations that hold positional information about the nucleotides in a sequence perform substantially better.

% [X]
\citeauthor{Gunasekaran2021AnalysisModels}~\cite{Gunasekaran2021AnalysisModels} used \gls{CNN}, \gls{CNN}-\gls{LSTM}, and \gls{CNN}-Bidirectional \gls{LSTM} architectures with Label and k-mer encoding for \gls{DNA} sequence classification. Different classification metrics were used to evaluate the models. According to the findings of the experiments, the \gls{CNN} and \gls{CNN}-Bidirectional \gls{LSTM} with k-mer encoding have good accuracy, with 93.16\% and 93.13\% on testing data, respectively.

% [X]
\citeauthor{Lugo2021AIdentification}~\cite{Lugo2021AIdentification} presented a sequential \gls{DL} approach for bacterium identification. To derive an identification model for whole-genome bacterium sequences, the proposed neural network takes advantage of the massive volumes of data supplied by Next-Generation Sequencing. The bidirectional \gls{RNN} (BI-\gls{GRU}) outperformed other classification algorithms (Naive Bayes, Multilayer Perceptron) after verifying the identification model. In a low-dimensional space, a distributed representation was proven as the appropriate encoding for bacterial genetic information. The distributed representation is given context by combining two or more k-mer lengths. Context makes use of positional data, which is critical in biological sequences.

Table~\ref{tab:previous_work} provides an overview of the previous work on DNA classification described above.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{table}[ht]
	\caption{Overview of DNA classification's previous work.}
	\label{tab:previous_work}
\centering
\scalebox{0.8}{
\begin{tabular}{lp{2cm}p{3.8cm}p{3.5cm}p{2.5cm}p{3.3cm}}
	\toprule
	\textbf{Year} & \textbf{Authors} & \textbf{Title} & \textbf{Focus} & \textbf{Classifier} & \textbf{Features / Encoding}\\
	\midrule
	
	\citeyear{Nair2010ANNRepresentation} & \citeauthor{Nair2010ANNRepresentation} & \citetitle{Nair2010ANNRepresentation}~\cite{Nair2010ANNRepresentation} & Taxonomic classification of Eukaryotic species & \gls{ANN} & FCGR\\\midrule
	
	\citeyear{Rizzo2015AClassification} & \citeauthor{Rizzo2015AClassification} & \citetitle{Rizzo2015AClassification}~\cite{Rizzo2015AClassification} & Classify 16S bacterial genomic sequences & \gls{DNN}, General Regression Neural Network, Naive Bayes, \gls{RF}, \gls{SVM} & k-mer encoding\\\midrule
	
	\citeyear{Nguyen2016DNANetwork} & \citeauthor{Nguyen2016DNANetwork} & \citetitle{Nguyen2016DNANetwork}~\cite{Nguyen2016DNANetwork} & Solve DNA sequence classification problem in 12 datasets & \gls{CNN} & one-hot vectors\\\midrule
	
	\citeyear{LoBosco2017DeepClassification} & \citeauthor{LoBosco2017DeepClassification} & \citetitle{LoBosco2017DeepClassification}~\cite{LoBosco2017DeepClassification} & Classification of bacteria species with no steps of sequence preprocessing & \gls{CNN}, \gls{RNN} & Character-level one-hot encoding\\\midrule
	
	\citeyear{Chen2017AClassification} & \citeauthor{Chen2017AClassification} & \citetitle{Chen2017AClassification}~\cite{Chen2017AClassification} & Cell-free DNA and normal DNA classification & \gls{KNN}, \gls{SVM}, \gls{RF} & DNA fragmentation patterns\\\midrule
	
	\citeyear{Helaly2019ConvolutionalStudy} & \citeauthor{Helaly2019ConvolutionalStudy} & \citetitle{Helaly2019ConvolutionalStudy}~\cite{Helaly2019ConvolutionalStudy} & 16S rRNA barcode dataset taxonomy classification & \gls{CNN} & k-mer spectral representation, \gls{FCGR}, and character-level integer encoding\\\midrule
	
	\citeyear{Gunasekaran2021AnalysisModels} & \citeauthor{Gunasekaran2021AnalysisModels} & \citetitle{Gunasekaran2021AnalysisModels}~\cite{Gunasekaran2021AnalysisModels} & COVID, SARS, MERS, dengue, hepatitis, and influenza classification & \gls{CNN}, \gls{CNN}-\gls{LSTM}, \gls{CNN}-Bidirectional, \gls{LSTM} & Label and k-mer encoding\\\midrule
	
    \citeyear{Lugo2021AIdentification} & \citeauthor{Lugo2021AIdentification} & \citetitle{Lugo2021AIdentification}~\cite{Lugo2021AIdentification} & Bacterium identification & BI-\gls{GRU} & Distributed k-mer \\
    
	\bottomrule
\end{tabular}
}
\end{table}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Tools for building DNA sequence classification algorithms}

Vale a pena falar outra vez disto se já tenho a secção~\ref{sec:ML_DL_libraries}?

\section{asdasd}

Understanding the connections between protein structure and function is one of biology's main goals. The basic amino acid sequences provide particularly helpful structural information for understanding the structure-function paradigm. The idea that sequences with similar structures have comparable functions is used to classify DNA sequences. Sequence alignment techniques such as BLAST~\cite{Altschul1990BasicTool} and FASTA~\cite{Pearson1988ImprovedComparison} have historically been used to determine sequence similarity. This decision is based on two primary assumptions: (1) the functional components have similar sequence properties, and (2) the functional elements' relative order is preserved across sequences. These assumptions are applicable in a wide range of situations, but they are not universal~\cite{LoBosco2017DeepClassification}.

Regardless, despite recent advancements, the major issue that severely restricts the use of alignment techniques remains their computational time complexity. As a result, alignment-free approaches~\cite{Vinga2003Alignment-freeReview,Pinello2014ApplicationsEpigenomics} that have recently been developed have emerged as a viable strategy to investigating the regulatory genome. Feature extraction, such as spectral representation of DNA sequences~\cite{LoBosco2014ASequences,LoBosco2015AlignmentClassification}, is undoubtedly a part of these approaches. Feature representations, which are often constructed by specialists, are critical to the effectiveness of traditional \gls{ML} algorithms. Following that, it is vital to choose which features are more suited for dealing with the particular problem, which is still a very important and challenging stage today~\cite{LoBosco2017DeepClassification}.